File: internal/hardware/hardware.go
// Package hardware reads BLIS's hardware catalogue so an authoring mistake — a
// name that does not exist, or two names for the same accelerator — is caught
// before a run rather than discovered as a duplicate row.
package hardware

import (
	"encoding/json"
	"fmt"
	"os"
	"sort"
	"strings"

	"github.com/inference-sim/leaderboard/internal/schema"
)

// DefaultConfigPath is where the catalogue lives relative to this repo's root.
const DefaultConfigPath = "../inference-sim/hardware_config.json"

// Catalog is the set of hardware names BLIS knows, with the numeric spec of each.
type Catalog struct {
	specs map[string]map[string]any
}

// Load reads hardware_config.json. Keys prefixed with "_comment" are prose, not
// hardware, and are dropped — including at the top level, where the trimmed
// testdata copy carries its own provenance note.
func Load(path string) (*Catalog, error) {
	raw, err := os.ReadFile(path)
	if err != nil {
		return nil, fmt.Errorf("hardware: read %s: %w", path, err)
	}
	var doc map[string]json.RawMessage
	if err := json.Unmarshal(raw, &doc); err != nil {
		return nil, fmt.Errorf("hardware: parse %s: %w", path, err)
	}

	c := &Catalog{specs: make(map[string]map[string]any, len(doc))}
	for name, body := range doc {
		if strings.HasPrefix(name, "_comment") {
			continue
		}
		var spec map[string]any
		if err := json.Unmarshal(body, &spec); err != nil {
			return nil, fmt.Errorf("hardware: parse %s entry %q: %w", path, name, err)
		}
		for k := range spec {
			if strings.HasPrefix(k, "_comment") {
				delete(spec, k)
			}
		}
		c.specs[name] = spec
	}
	if len(c.specs) == 0 {
		return nil, fmt.Errorf("hardware: %s contains no hardware entries", path)
	}
	return c, nil
}

// Known reports whether name is a hardware entry BLIS would accept.
func (c *Catalog) Known(name string) bool {
	_, ok := c.specs[name]
	return ok
}

// Names returns every hardware name, sorted.
func (c *Catalog) Names() []string {
	out := make([]string, 0, len(c.specs))
	for name := range c.specs {
		out = append(out, name)
	}
	sort.Strings(out)
	return out
}

// Aliases returns the other names whose numeric spec is identical to name's. Two
// such names are one accelerator, so running both produces a duplicate row rather
// than a second candidate.
func (c *Catalog) Aliases(name string) []string {
	spec, ok := c.specs[name]
	if !ok {
		return nil
	}
	key, err := schema.CanonicalJSON(spec)
	if err != nil {
		return nil
	}

	var out []string
	for other, otherSpec := range c.specs {
		if other == name {
			continue
		}
		otherKey, err := schema.CanonicalJSON(otherSpec)
		if err != nil {
			continue
		}
		if string(key) == string(otherKey) {
			out = append(out, other)
		}
	}
	sort.Strings(out)
	return out
}

// Entry is one accelerator the way the Catalog page shows it: its name, blis's
// numeric spec, and the other names whose spec is byte-identical (aliases — one
// accelerator, not a rival candidate).
type Entry struct {
	Name    string         `json:"name"`
	Aliases []string       `json:"aliases"`
	Spec    map[string]any `json:"spec"`
}

// Entries returns every hardware entry, sorted by name, exposing the parsed specs
// the Catalog page renders. It reuses Names and Aliases so the sort order and the
// alias grouping cannot drift from what the run-time checks already use.
func (c *Catalog) Entries() []Entry {
	names := c.Names()
	out := make([]Entry, 0, len(names))
	for _, name := range names {
		// Aliases returns nil for an accelerator with none; send [] rather than a nil slice
		// so the JSON is "aliases":[] not null, and the client never has to guard a spread.
		aliases := c.Aliases(name)
		if aliases == nil {
			aliases = []string{}
		}
		out = append(out, Entry{
			Name:    name,
			Aliases: aliases,
			Spec:    c.specs[name],
		})
	}
	return out
}

File: web/src/hardware.ts
/**
 * The accelerator catalog the Catalog page shows, fetched live from the server rather
 * than frozen into the front-end. `leaderboard serve` reads it from the upstream
 * hardware_config.json (GET /api/hardware) the same file blis checks a run's hardware
 * against, so the list is never a hand-maintained snapshot. Kept out of the components,
 * like models.ts and workloads.ts, so every claim it makes is testable.
 *
 * There is no committed fallback list: when the config cannot be read (no server, or a
 * missing hardware_config.json) the page says so, exactly as the Workloads tab reports an
 * unreachable server.
 */

/** One accelerator: its name, the spec-identical names that are the same accelerator
 * (aliases, not rival candidates), and blis's numeric spec. Mirrors internal/hardware.Entry. */
export interface HardwareInfo {
  name: string
  aliases: string[]
  spec: Record<string, number>
}

const UNREACHABLE =
  'Could not reach the hardware server. Start it with `make build && ./bin/leaderboard serve`, then try again.'

/**
 * The other names of `name`'s accelerator: the spec-identical names, which are the same
 * accelerator, not rival candidates. The server hands each entry its own aliases, so this
 * answers the question whichever name is asked (a canonical name or one of its aliases). An
 * unknown name has none. The Declare form's duplicate-row guard reads this to refuse a
 * candidate whose accelerator already sits in the table under one of these names.
 */
export function hardwareAliases(hardware: HardwareInfo[], name: string): string[] {
  const entry = hardware.find((h) => h.name === name || (h.aliases ?? []).includes(name))
  if (!entry) return []
  return [entry.name, ...(entry.aliases ?? [])].filter((n) => n !== name).sort()
}

/**
 * Folds spec-identical names into one entry each, so the same accelerator is offered once
 * rather than as two rival candidates (running both would be a duplicate row, not a second
 * choice). The server sends every name with the others as its aliases; each group keeps its
 * more specific name (the longer one, ties broken alphabetically) as canonical and lists the
 * rest as aliases. The choice is cosmetic — blis treats the names as interchangeable — but it
 * keeps the label stable (A100-SXM over A100-80). Entries are returned sorted by name.
 */
export function collapseHardwareAliases(hardware: HardwareInfo[]): HardwareInfo[] {
  const seen = new Set<string>()
  const out: HardwareInfo[] = []
  for (const h of hardware) {
    if (seen.has(h.name)) continue
    const group = [h.name, ...(h.aliases ?? [])]
    for (const n of group) seen.add(n)
    // Prefer the longer name (more specific), and among equal lengths the alphabetically
    // first, so the canonical name is deterministic. group always holds h.name, so the
    // fallback only satisfies the type and never triggers.
    const canonical =
      [...group].sort((a, b) => b.length - a.length || a.localeCompare(b))[0] ?? h.name
    out.push({
      name: canonical,
      aliases: group.filter((n) => n !== canonical).sort(),
      spec: h.spec,
    })
  }
  return out.sort((a, b) => a.name.localeCompare(b.name))
}

/** listHardware fetches the catalog. A fetch rejection is "the server is not running"; an
 * HTTP error carries the server's {error} message. fetchImpl is injectable so the call is
 * unit-tested without a server, the same shape as models.listModels. */
export async function listHardware(fetchImpl: typeof fetch = fetch): Promise<HardwareInfo[]> {
  let res: Response
  try {
    res = await fetchImpl('/api/hardware')
  } catch {
    throw new Error(UNREACHABLE)
  }
  const text = await res.text()
  let parsed: unknown = null
  try {
    parsed = text ? JSON.parse(text) : null
  } catch {
    parsed = null
  }
  if (!res.ok) {
    const message =
      parsed && typeof parsed === 'object' && 'error' in parsed
        ? String((parsed as { error: unknown }).error)
        : `The hardware server returned HTTP ${res.status}.`
    throw new Error(message)
  }
  return (parsed as { hardware?: HardwareInfo[] } | null)?.hardware ?? []
}
